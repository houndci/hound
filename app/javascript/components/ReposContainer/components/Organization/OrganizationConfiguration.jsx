import React from "react";
import classNames from "classnames";
import $ from "jquery";

import "object-assign-shim";

import * as Ajax from "../../../../modules/Ajax";

export default class OrganizationConfiguration extends React.Component {
  constructor(props) {
    super(props);

    let repo = "";

    if (props.repo) {
      repo = props.repo;
    }

    this.state = {
      configUpdated: false,
      enabled: !!props.enabled,
      repo
    };
  }

  setEnabled(enabled) {
    this.setState({ enabled });
    this.updateOwner({ config_enabled: enabled });
  }

  setRepo(repo) {
    this.setState({ repo });
    this.updateOwner({ config_repo: repo });
  }

  handleChange(event) {
    this.setRepo(event.target.value);
  }

  toggle() {
    const enabled = !this.state.enabled;
    this.setEnabled(enabled);
  }

  updateOwner(options) {
    const currentState = {
      config_enabled: this.state.enabled,
      config_repo: this.state.repo
    };

    this.setState({ configUpdated: false });
    Ajax.updateOwner(this.props.id, { ...currentState, ...options }).then(() =>
      this.setState({ configUpdated: true })
    );
  }

  renderRepoSelect() {
    if (this.state.enabled) {
      return (
        <div className="organization-header-source">
          <span className="organization-header-label">Use .hound.yml from</span>
          <select
            className="organization-header-select"
            onChange={event => this.handleChange(event)}
            value={this.state.repo}
          >
            {this.renderRepoOptions()}
          </select>
          <div className="inline-flash--success config-enabled">
            {this.state.configUpdated &&
              <span data-role="config-saved">&#10004;</span>}
          </div>
        </div>
      );
    } else {
      return null;
    }
  }

  renderRepoOptions() {
    return this.props.repos.map(repo =>
      <option key={repo.id} value={repo.name}>
        {repo.name}
      </option>
    );
  }

  render() {
    const switchId = `toggle-${this.props.id}`;
    return (
      <div className="organization-header-config">
        <div className="organization-header-toggle">
          <span className="organization-header-label">
            Use organization-wide config
          </span>

          <input
            id={switchId}
            checked={this.state.enabled}
            className="organization-header-toggle-input"
            onChange={() => this.toggle()}
            type="checkbox"
            name="toggle"
          />

          <label className="toggle-switch" htmlFor={switchId} />
        </div>

        {this.renderRepoSelect()}
      </div>
    );
  }
}
