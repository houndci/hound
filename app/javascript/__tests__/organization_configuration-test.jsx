import OrganizationConfiguration from "../components/ReposContainer/components/Organization/OrganizationConfiguration";

const ownerId = 1;
const repos = [
  {
    id: 1,
    name: "Test repo",
    owner: {
      id: 1
    }
  },
  {
    id: 2,
    name: "Test repo 2",
    owner: {
      id: 1
    }
  }
];

describe("with a user that is not an organization owner", () => {
  const userIsOrgOwner = false;

  describe("and when config is disabled", () => {
    it("renders a config is disabled message", () => {
      const configEnabled = false;
      const configRepo = null;

      const wrapper = shallow(
        <OrganizationConfiguration
          enabled={configEnabled}
          id={ownerId}
          repo={configRepo}
          repos={repos}
          userIsOrgOwner={userIsOrgOwner}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("and when config is enabled but a repo has not been set", () => {
    it("renders a config is disabled message", () => {
      const configEnabled = true;
      const configRepo = null;

      const wrapper = shallow(
        <OrganizationConfiguration
          enabled={configEnabled}
          id={ownerId}
          repo={configRepo}
          repos={repos}
          userIsOrgOwner={userIsOrgOwner}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("and when config is enabled and a config repo is set", () => {
    it("renders a config is enabled message", () => {
      const configEnabled = true;
      const configRepo = 'Test repo 2';

      const wrapper = shallow(
        <OrganizationConfiguration
          enabled={configEnabled}
          id={ownerId}
          repo={configRepo}
          repos={repos}
          userIsOrgOwner={userIsOrgOwner}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});

describe("with a user that is an organization owner", () => {
  const userIsOrgOwner = true;

  describe("and when config is disabled", () => {
    it("renders the appropriate config options", () => {
      const configEnabled = false;
      const configRepo = null;

      const wrapper = shallow(
        <OrganizationConfiguration
          enabled={configEnabled}
          id={ownerId}
          repo={configRepo}
          repos={repos}
          userIsOrgOwner={userIsOrgOwner}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("and when config is enabled but a repo has not been set", () => {
    it("renders the appropriate config options", () => {
      const configEnabled = true;
      const configRepo = null;

      const wrapper = shallow(
        <OrganizationConfiguration
          enabled={configEnabled}
          id={ownerId}
          repo={configRepo}
          repos={repos}
          userIsOrgOwner={userIsOrgOwner}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("and when config is enabled and a config repo is set", () => {
    it("renders the appropriate config options", () => {
      const configEnabled = true;
      const configRepo = 'Test repo 2';

      const wrapper = shallow(
        <OrganizationConfiguration
          enabled={configEnabled}
          id={ownerId}
          repo={configRepo}
          repos={repos}
          userIsOrgOwner={userIsOrgOwner}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
