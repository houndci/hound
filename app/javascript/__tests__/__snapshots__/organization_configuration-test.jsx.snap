exports[`with a user that is an organization owner and when config is disabled renders the appropriate config options 1`] = `
<div
  className="organization-header-config">
  <div
    className="organization-header-toggle">
    <span
      className="organization-header-label">
      Use organization-wide config
    </span>
    <input
      checked={false}
      className="organization-header-toggle-input"
      id="toggle-1"
      name="toggle"
      onChange={[Function]}
      type="checkbox" />
    <label
      className="toggle-switch"
      htmlFor="toggle-1" />
  </div>
</div>
`;

exports[`with a user that is an organization owner and when config is enabled and a config repo is set renders the appropriate config options 1`] = `
<div
  className="organization-header-config">
  <div
    className="organization-header-toggle">
    <span
      className="organization-header-label">
      Use organization-wide config
    </span>
    <input
      checked={true}
      className="organization-header-toggle-input"
      id="toggle-1"
      name="toggle"
      onChange={[Function]}
      type="checkbox" />
    <label
      className="toggle-switch"
      htmlFor="toggle-1" />
  </div>
  <div
    className="organization-header-source">
    <span
      className="organization-header-label">
      Use .hound.yml from
    </span>
    <select
      className="organization-header-select"
      onChange={[Function]}
      value="Test repo 2">
      <option>
        None
      </option>
      <option
        value="Test repo">
        Test repo
      </option>
      <option
        value="Test repo 2">
        Test repo 2
      </option>
    </select>
    <div
      className="inline-flash--success config-enabled" />
  </div>
</div>
`;

exports[`with a user that is an organization owner and when config is enabled but a repo has not been set renders the appropriate config options 1`] = `
<div
  className="organization-header-config">
  <div
    className="organization-header-toggle">
    <span
      className="organization-header-label">
      Use organization-wide config
    </span>
    <input
      checked={true}
      className="organization-header-toggle-input"
      id="toggle-1"
      name="toggle"
      onChange={[Function]}
      type="checkbox" />
    <label
      className="toggle-switch"
      htmlFor="toggle-1" />
  </div>
  <div
    className="organization-header-source">
    <span
      className="organization-header-label">
      Use .hound.yml from
    </span>
    <select
      className="organization-header-select"
      onChange={[Function]}
      value="">
      <option>
        None
      </option>
      <option
        value="Test repo">
        Test repo
      </option>
      <option
        value="Test repo 2">
        Test repo 2
      </option>
    </select>
    <div
      className="inline-flash--success config-enabled" />
  </div>
</div>
`;

exports[`with a user that is not an organization owner and when config is disabled renders a config is disabled message 1`] = `
<div
  className="organization-header-config">
  <div
    className="organization-header-toggle">
    <span
      className="organization-header-label">
      Organization-wide config is disabled (only Organization owners can activate)
    </span>
  </div>
</div>
`;

exports[`with a user that is not an organization owner and when config is enabled and a config repo is set renders a config is enabled message 1`] = `
<div
  className="organization-header-config">
  <div
    className="organization-header-toggle">
    <span
      className="organization-header-label">
      Using organization-wide config from Test repo 2
    </span>
  </div>
</div>
`;

exports[`with a user that is not an organization owner and when config is enabled but a repo has not been set renders a config is disabled message 1`] = `
<div
  className="organization-header-config">
  <div
    className="organization-header-toggle">
    <span
      className="organization-header-label">
      Organization-wide config is disabled (only Organization owners can activate)
    </span>
  </div>
</div>
`;
